### PR TITLE
solution

### PR DESCRIPTION
**Pull request by olena-yanovska**
[DEMO LINK](https://olena-yanovska.github.io/react_todo-app-with-api/)

### CODE REVIEW

**Comment by NadiiaKoch** [comment]
Знімок екрана 2023-01-31 о 15 21 12

File: src/components/Footer/Footer.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        onClick={() => setCompletedFilter(CompletedFilter.Completed)}
+      >
+        Completed
+      </a>
+    </nav>
+
+    <button
+      data-cy="ClearCompletedButton"
+      type="button"
+      className="todoapp__clear-completed"
```
Comment (NadiiaKoch) [inline src/components/Footer/Footer.tsx:65]: Add 'is-invisible' to className when we have at least one completed todo.

**Review by NadiiaKoch — CHANGES_REQUESTED** [review]
Will be good to fix bug with title

File: src/components/Footer/Footer.tsx (hunk 1 above)
Comment (olena-yanovska) [reply src/components/Footer/Footer.tsx]: done

**Comment by olena-yanovska** [comment]
> Will be good to fix bug with title

thanks, fixed it

**Review by NadiiaKoch — APPROVED** [review]
GJ!

File: src/components/TodoList/TodoList.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}) => {
+  return (
+    <section className="todoapp__main" data-cy="TodoList">
+      {todos.map((todo => (
+        <TodoItem
+          todo={todo}
+          key={todo.id}
+          deleteTodo={deleteTodo}
+          isDeleting={deletingTodoIds.includes(todo.id)}
+          onChangeTodoStatus={changeTodoStatus}
```
Comment (ivan-kadykalo) [inline src/components/TodoList/TodoList.tsx:36]:
```suggestion
          changeTodoStatus={changeTodoStatus}
```

File: src/components/TodoList/TodoList.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+          newTodoField={newTodoField}
+        />
+      )))}
+
+      {tempTodo && (
+        <TodoItem
+          todo={tempTodo}
+          deleteTodo={deleteTodo}
+          isDeleting={deletingTodoIds.includes(tempTodo.id)}
+          onChangeTodoStatus={changeTodoStatus}
```
Comment (ivan-kadykalo) [inline src/components/TodoList/TodoList.tsx:47]:
```suggestion
          changeTodoStatus={changeTodoStatus}
```

File: src/components/TodoItem/TodoItem.tsx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+  newTodoField: React.RefObject<HTMLInputElement>;
+  isDeleting: boolean,
+  isAdding?: boolean;
+}
+
+export const TodoItem: React.FC<Props> = memo(({
+  todo,
+  deleteTodo,
+  onChangeTodoStatus,
```
Comment (ivan-kadykalo) [inline src/components/TodoItem/TodoItem.tsx:23]:
```suggestion
  changeTodoStatus,
```

File: src/components/TodoItem/TodoItem.tsx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  onUpdateTodo,
```
Comment (ivan-kadykalo) [inline src/components/TodoItem/TodoItem.tsx:24]:
```suggestion
  updateTodo,
```

File: src/components/Header/Header.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  showError: (message: string) => void;
+  onChangeAllTodos: () => void;
+  isAllTodosCompleted: boolean;
+  isAddingTodo: boolean,
+}
+
+export const Header: React.FC<Props> = memo(({
+  onAddTodo,
+  showError,
+  onChangeAllTodos,
```
Comment (ivan-kadykalo) [inline src/components/Header/Header.tsx:24]: or `handleChangeAllTodos`
```suggestion
  changeAllTodos,
```

**Review by ivan-kadykalo — CHANGES_REQUESTED** [review]
When you update todo (rename title or change completed status) there should be a loading spinner too, as for adding or deleting.

File: src/App.tsx [hunk 7]
```diff
@@ -1,223 +1,192 @@
...
   const newTodoField = useRef<HTMLInputElement>(null);
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+    if (user) {
+      todoApi.getTodos(user.id)
+        .then(setTodos)
+        .catch(() => showError('Todos loading failed'));
```
[the remark this answers is not in the document]
Comment (olena-yanovska) [reply src/App.tsx:42]: done

**Review by ivan-kadykalo — APPROVED** [review]

File: src/App.tsx [hunk 8]
```diff
@@ -1,223 +1,192 @@
...
+import { useError } from './controllers/useErrors';
 
 export const App: React.FC = () => {
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [completedFilter, setCompletedFilter] = useState(CompletedFilter.All);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isAddingTodo, setIsAddingTodo] = useState(false);
+  const [deletingTodoIds, setDeletingTodoIds] = useState<number[]>([]);
+  const [, setSelectedTodoIds] = useState<number[]>([]);
```
[the remark this answers is not in the document]
Comment (olena-yanovska) [reply src/App.tsx]: I wanted to use it for Loader component, fixed

**Pull request closed without merging**